### DISCUSSION TITLE
Private state and automatic store disposal in options stores

### DISCUSSION THREAD

**Discussion by ilyarefraction** [vuejs/pinia#2637] [Help and Questions] [answered]
Recently I encountered a case when I needed a private state in the `options` API store. While searching for an answer I found it in this [article ](https://masteringpinia.com/blog/my-top-5-tips-for-using-pinia#:~:text=Private%20state%20with%20nested%20Stores)

This works like a charm in `setup` stores but doesn't fully resolve the issue when using `options` stores, i.e.
```js
// store.js
import { defineStore } from 'pinia';

export const usePrivate = defineStore('auth-private', () => {
  const token = null;
  return {
    token,
  };
});

const myStore = defineStore('my-store', {
  state: () => ({
    name: 'Eduardo',
  }),
  getters: {
    privateStore() {
      return usePrivate();
    },
  },
  actions: {
    getToken() {
      if (this.privateStore.token) {
        return this.privateStore.token;
      }
      return 'no token';
    },
  },
})();

const TOKEN = myStore.privateStore.token;
```
Here I see two issues:
1. Anyone can access the data from the `privateStore`, so this is not a **private** data
2. If I want to delete this *private* store I have to do it outside of the store i.e.:
```js
import myStore from './myStore';

onBeforeUnmount(() => {
  myStore.$dispose();
  myStore.privateStore.$dispose();
});
```
This looks a bit funny to me as someone (I mean other components) always should know about the store's private state.

------
I'll be very happy if these issues can be somehow resolved in `options` stores, but I don't think this is possible with the current pinia API.

Of course, this is only a limitation of the `options` API stores and can be easily resolved just by using `setup` stores, but this leads to an inconsistency in the codebase as `options` store syntax is [suggested by default in this `pinia` docs](https://pinia.vuejs.org/core-concepts/#What-syntax-should-I-pick-). 

I suggest:
* A more detailed guide on how to choose the store syntax and possible `options` stores limitations.

**Comment by posva** [comment]
FYI: options stores are not the default. You can use both in the same project (mentioned in docs). They cater to different ways of organization, but due to the nature of the setup stores, they will always be more flexible than option stores.

In your case, the private store is accessible because you return it in a getter. To keep it private, you need to call `useStore()` in every action and getter that needs it. If that repetition bothers you, I recommend you to switch over to setup stores.

**Reply by ilyarefraction** [reply]
Thanks for your reply!

1. > FYI: options stores are not the default. You can use both in the same project (mentioned in docs)

I understand that options stores are not "default", I mean that they are **suggested** to try first. By this discussion, I only wanted to highlight to you some possible docs problems.

For example, our team started using different stores api only **after** the discovery of these options api store problems, which resulted in some time spent rewriting some options api stores to setup syntax. Maybe if some options stores drawbacks were described more clearly in [What syntax should I pick](https://pinia.vuejs.org/core-concepts/#What-syntax-should-I-pick-) (or any other related section) we would start by using the setup stores.

2.  >  To keep it private, you need to call useStore() in every action and getter that needs it.
Thanks, I saw this in docs but dropped this method as:
- This repetition indeed bothers us :)
- In our app, we are currently disposing and deleting unused stores. We find it useful as at least debugging is made easier as dev tools are not polluted with unrelated stores. I didn't find a way to delete and dispose store if we call `useStore()` in every action and getter.

Feel free to close this discussion at any time if you think this is unrelated or if you have any other reasons.

**Reply by posva** [reply]
I do want to update the docs to better show the different syntaxes but I lack the funding (by companies) to spend time on it 🫠

>  I didn't find a way to delete and dispose store if we call useStore() in every action and getter.

It shouldn't be a problem as long as you dispose both stores

**Reply by ilyarefraction** [reply]
> I do want to update the docs to better show the different syntaxes but I lack the funding (by companies) to spend time on it

Ok, that's reasonable, in any case I respect your time and not forcing you to do anything 

> It shouldn't be a problem as long as you dispose both stores
Sorry, I didn't quite understand

Can you check this [demo](https://play.pinia.vuejs.org/#eNqNVF1P2zAU/StWNCmt1NpDTEwqKSrbeBgPGxq8LXsIiVsMjm3ZTiiq8t93/ZE0pQiBhBrb55577rm+3iWXSuG2ockiyUypmbLIUNsoxAuxWeaJNXlykQtWK6kt2qENtZelZS29YYIVqENrLWuUKrdKR7j65dZKTQcAJsatsTWAykUphYFMHrLswZPp/sgTwtFhPofoQwvrQj0O+xVuC97QPUVFObU0yFiiyRQtL9AuFyikxZ8qZpQ0PiuK4GO6vxHMqn/nuegce0aCUWALLCytFQc4rBDKKtZeXDeQ3EoEVFCHZq1T6mkWaLeL2aGuO/lEBcjquoy4OEcwkNw6CRHv4seg7L6xVgq0Kjkrn6BJo0qhW+hHqCW4m5GAhsiMjMQmsyQ0a14XCj8aKeAKeHvyeACdBwFux+15Z9xOnpDgEjX1/F7LZ0M1xOfJrIeu4D6RirZWSm7mhWIh7MFaZRaElJUAPGhmrcaCWiJUTY5iVmf4DJ8Qzu4JJCJMVHR7mAYi5hWtP8LeQ1ef8ckX/NWztqeRtHasjhS624Et1sD9WbPNK1NKWSvGqf6tLIP7dWBOwbl8vvZ7Vjd0kFg+0PLpjf1Hsw2qbzQF+1po23BmCw2XIxxf3f6iW/geDmtZNRzQ7xz+oUbyxmkMsG+NqED2COfV/vQ9ZmJzZ662lgrTF+WEejc83tv8/Z3S93JP8enIxX7cwcDhUajomok4kYcvx3muYZrgn249NkxwY+hNnJ/lOHiSFo19mMfZSmfDcAMBirHWTReEiYZzz46QhodNix6GAmQWVp3HdNO3hfSv2SsV9cvclwkKIqkf1kXUMxkyiaKG3fSqagpdybTPOY0fBbxwkMe5GgNGz8OwhxBbo8nek8kU+woOIEOVx7jogvvr9p8RngoZ/Ait8KBepv+N1iTdf5OtGFY=)?

1. Open demo - see that initially, both stores are present
2. Click the 'Delete store' button
3. See only 'my-store' was deleted and 'auth-private' is still in state

My question is - how do I delete the private store using the options store syntax in this case? I can't access the private store outside of the file where it is defined so I need some way to delete store in inside it (sadly, I can't use `onScopeDispose` api in options stores).

The only way I found is to define a pinia plugin which will handle store deletion and add an extra action that should be called when the store is deleted. [Demo](https://play.pinia.vuejs.org/#eNqNVE1v2zAM/SuEMSAOkEgrOnRAmwTpth7Ww1asvc07uLaSqrUlQ5LdFEH++6gPf6VdsUMQS3x8JJ9I7qPLqiJNzaLzaKEzxSsDmpm6giIV22USGZ1Eq0TwspLKwB62zFxmhjfshgueziBTLDX+AAfYKFnCpLKnycCpfLk1UrEOQKi2Z2I0ohKRSaENpD0tLI/ixNMx6tZgUEQNfIi2d6RJi5pZ0qGp1iyO9+CCnoNLLyQ0heUK9okA8PR7+JBzXUltk10OoBeWE0ChNkp4D0AsK5jBbBxv3JN5oyeKp+jrrzz8X1n/7qORDzz/E9wOM/t/wNPBMrVCuGIwxaCtVYhSuK7RZCQgk4FK8cbq5KAJOllmFPZOPjFhHVquQRnIOKgDGddGz/lWOAbwVGRYt+Vp7/31jQ/bWQ820IL67sJewoNhZVUgZgVgvRc5b1buTc9hvx+98eGwoNbqYPe1MVLAOit49oTNOUgDuxS+eXm9Mgvq0ei5oF28RESzyPflvEwr8qilwNZ3xSbBgB2PaXjxk8g9ir1JIuq+CdPl/F7JZ80U+ieRex8LXeMc0Zw1RspCz9OKe7cHYyp9TmmWC8RjzrxRRDBDRVXSVz7rM3JGTmjB7ykGolzkbDcOgx7znJX/w95C1x/JySfy2bE2p4G0tKyut/CNUBajsRs2fHskSibLihdM/awMx24ZiZMWhXy+dndG1axLMXtg2dMb949657O+UQzla/DZOptJFfamN1/d/mA7/O6MpczrAtHvGH8xLYva5uhhX2qRY9oDnMv2u3tjLrZ3+mpnmNBtUTZRp4bDO5m/vlN6n+4pOR2o2G42FLDbfznbcBHma7wkLxKFA4E/tnNYP4+4r8IQ4TgOnONJWpuHeZjryawbVSRoN5ixw41uoi4Kxz5YWu4EHjLzJ9wr+G83y1uJtIv7KIvyZe7KxAwCqVtj7QqMu0giLfF2cpXXqcrlpI05DR920DGOVTU49NupvwPgG4h7TeIpcRWMIF2Vr3FBBbdL+88Anwjp9fBPERZu+Hprnw2Djtdjdz1K4WhXtjGCEi5UUD86/AVFNIvQ)

Maybe there is a more elegant way to do this?

**Reply by posva** [reply, marked as answer]
You have many ways. The easiest way is to export a function to dispose the store. The plugin approach should also work

**Comment by martishkaass** [comment]
Lol 🤡
![1dd38e839851ace95c81a6d18ecd8a89](https://github.com/vuejs/pinia/assets/75129220/315562ba-f9f6-418c-b915-1c2c83dd4b9c)
